File: restdoc-rpc-client/restdoc-restweb-client/src/main/kotlin/restdoc/client/restweb/RestWebAgentClientConfiguration.java
package restdoc.client.restweb;

import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.core.PriorityOrdered;
import org.springframework.core.env.Environment;
import org.springframework.web.client.RestTemplate;
import restdoc.client.api.Agent;
import restdoc.client.api.AgentClientConfiguration;
import restdoc.client.api.AgentConfigurationProperties;
import restdoc.client.api.AgentImpl;
import restdoc.client.restweb.context.EndpointsListener;
import restdoc.client.restweb.handler.ExportApiHandler;
import restdoc.client.restweb.handler.InvokerApiHandler;
import restdoc.client.restweb.handler.ReportClientInfoHandler;
import restdoc.remoting.netty.NettyRequestProcessor;

/**
 * @author Maple
 */
@Configuration
@Import(value = {EnvConfiguration.class})
public class RestWebAgentClientConfiguration implements AgentClientConfiguration, PriorityOrdered {

    private final AgentImpl agentImpl;

    private final ReportClientInfoHandler reportClientInfoHandler;

    private final ExportApiHandler exportApiHandler;

    private final InvokerApiHandler invokerApiHandler;

    public RestWebAgentClientConfiguration(@Qualifier(value = "restWebAgentImpl") AgentImpl agentImpl,
                                           ReportClientInfoHandler reportClientInfoHandler,
                                           ExportApiHandler exportApiHandler,
                                           InvokerApiHandler invokerApiHandler) {
        this.agentImpl = agentImpl;
        this.reportClientInfoHandler = reportClientInfoHandler;
        this.exportApiHandler = exportApiHandler;
        this.invokerApiHandler = invokerApiHandler;
    }

    @Bean
    @ConditionalOnMissingBean
    public EndpointsListener endpointsListener(Environment environment) {
        return new EndpointsListener(environment);
    }

    @Bean
    @ConditionalOnMissingBean
    public InvokerApiHandler invokerApiHandler(RestWebInvokerImpl invoker) {
        return new InvokerApiHandler(invoker);
    }

    @Bean
    @ConditionalOnMissingBean
    public RestWebInvokerImpl restWebInvoker(Environment environment, RestTemplate restTemplate) {
        return new RestWebInvokerImpl(environment, restTemplate);
    }

    @Bean
    @ConditionalOnMissingBean
    public RestTemplate restTemplate(){return new RestTemplate();}

    @Bean
    @ConditionalOnMissingBean
    public ReportClientInfoHandler reportClientInfoHandler(AgentConfigurationProperties configurationProperties,
                                                           Environment environment) {
        return new ReportClientInfoHandler(configurationProperties, environment);
    }

    @Override
    public int getOrder() {
        return 0;
    }

    @NotNull
    @Override
    public NettyRequestProcessor getInvokeAPIHandler() {
        return this.invokerApiHandler;
    }

    @NotNull
    @Override
    public NettyRequestProcessor getReportClientInfoHandler() {
        return this.reportClientInfoHandler;
    }

    @NotNull
    @Override
    public NettyRequestProcessor getExportAPIHandler() {
        return this.exportApiHandler;
    }

    @NotNull
    @Override
    public Agent getAgent() {
        return this.agentImpl;
    }

   /* @Override
    public void registryRemotingTask() {

    }

    @Override
    public void registryRemotingHandler() {

    }*/

    @Override
    public void run(@NotNull String... args) {

        Agent agent = this.getAgent();

        // 1 registryRemotingTask
        this.registryRemotingTask();

        // 2 registryRemotingHandler
        this.registryRemotingHandler();

        // 3 start agent
        agent.start();
    }
}

File: restdoc-rpc-client/restdoc-restweb-client/src/main/java/restdoc/client/restweb/RestWebAgentClientConfiguration.kt
package restdoc.client.restweb

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import restdoc.client.api.*
import restdoc.client.restweb.context.EndpointsListener
import restdoc.remoting.InvokeCallback
import restdoc.remoting.common.ApplicationType
import restdoc.remoting.common.RemotingUtil
import restdoc.remoting.common.RequestCode
import restdoc.remoting.common.body.ClientInfoBody
import restdoc.remoting.common.body.RestWebExposedAPIBody
import restdoc.remoting.protocol.RemotingCommand

/**
 * @author Overman
 * @since 2020/9/15
 */
@Configuration
@Import(AgentConfiguration::class)
@ConditionalOnClass(value = [AgentConfiguration::class, EnvConfiguration::class])
open class RestWebAgentClientConfiguration : AgentClientConfiguration {

    @Autowired
    lateinit var agentConfigurationProperties: AgentConfigurationProperties

    @Autowired
    lateinit var agentImpl: AgentImpl

    @Autowired
    lateinit var endpointsListener: EndpointsListener

    override fun getAgent() = this.agentImpl

    /**
     *
     */
    override fun hook(): AgentHook {
        return object : AgentHook {
            override fun beforeStart() = listOf<AgentCallback>()
            override fun afterStart() = listOf<AgentCallback>()
        }
    }

    /**
     * registryRemotingTask
     */
    override fun registryRemotingTask() {
        // 1 Generate report client info interfaces
        agentImpl.addTask(reportClientInfoTask())

        // 2 Generate report http application exposed interfaces
        agentImpl.addTask(reportExposedInterfacesTask())
    }

    /**
     * Request process handelr
     */
    override fun registryRemotingHandler() {
    }


    private fun reportClientInfoTask(): RemotingTask {
        val serializationProtocol = "http"

        val body = ClientInfoBody()
        body.osname = System.getProperty("os.name")
        body.hostname = RemotingUtil.getHostname()
        body.service = agentConfigurationProperties.service
        body.applicationType = ApplicationType.DUBBO
        body.serializationProtocol = serializationProtocol

        val request = RemotingCommand.createRequestCommand(RequestCode.REPORT_CLIENT_INFO, null)
        request.body = body.encode()

        return RemotingTask(
                taskId = reportClientInfoTask,
                type = RemotingTaskType.SYNC,
                request = request,
                timeoutMills = 1000000L,
                invokeCallback = InvokeCallback { })
    }

    private fun reportExposedInterfacesTask(): RemotingTask {
        val body = RestWebExposedAPIBody()
        body.apiList = endpointsListener.restWebExposedAPIList
        val request = RemotingCommand.createRequestCommand(RequestCode.REPORT_EXPOSED_API, null)

        return RemotingTask(
                taskId = reportExposedInterfacesTask,
                type = RemotingTaskType.SYNC,
                request = request,
                timeoutMills = 1000000L,
                invokeCallback = InvokeCallback { })
    }
}